### DISCUSSION TITLE
How to Run Tomcat/Adempiere 3.9.4 ZK from Eclipse

### DISCUSSION THREAD

**Discussion by marcalwestf** [adempiere/adempiere#4058] [Show and tell]
The way used in ADempiere 3.9.3 to configure Tomcat doesn't work anymore. See issue https://github.com/adempiere/adempiere/issues/4016.

With version 3.9.4 of ADempiere, the application servers aren't embedded anymore in the installation; they are instead installed separately outside ADempiere and have only the configuration left inside ADempiere.

Here is a detailed explanation of a way of configuring a Tomcat server in Eclipse so it can be run or debugged inside Eclipse. 
I used 
- **Eclipse 2022-06** and 
- **Ubuntu 18.04**

I haven't managed (tried actually) to do the same with **Jetty** and **Wildfly**.

![imagen](https://user-images.githubusercontent.com/1789408/215533851-7fe57c0b-3e09-4eb3-b30d-b1ee599613a8.png)

**Step 1**
Download and Install **Tomcat 9** outside ADempiere:
![imagen](https://user-images.githubusercontent.com/1789408/215542884-0588f074-568b-42f8-b8e0-7ec3cdd709d7.png)

**Step 2** 
In Eclipse: Compile, Clean, and Build ADempiere
![imagen](https://user-images.githubusercontent.com/1789408/215545426-f2d36be9-6fbe-4bfa-82c1-7217e7f5d05a.png)

**Step 3**
Install Adempiere
(**this may be worth another discussion entry**)

**Step 4**
In Eclipse, select the main menu _Window/Preferences_ and make sure you have **JDK 11** installed:
![imagen](https://user-images.githubusercontent.com/1789408/215820763-dce42c9b-6d3f-4bfd-b6a5-41270956f06d.png)

Then, click on the project, open _context menu/Properties_, and make sure you have **JDK 11** set as the "compliant version" for the project.
![imagen](https://user-images.githubusercontent.com/1789408/215819826-010ed3e1-9e56-4af9-aa21-2eff7a210002.png)

Create a new _Server Configuration_. 
Make sure it is _Apache Tomcat 9 Server_, and it is the _localhost_.
Define a sensible name.
![Selection_495](https://user-images.githubusercontent.com/1789408/215545800-8181a95b-d06a-4a57-af42-272423701f8a.png)
![Selection_496](https://user-images.githubusercontent.com/1789408/215545821-6bbf0c6e-29c7-40c9-b9ba-634973e42351.png)

**Step 5**
Copy all files from ADempiere installation $ADEMPIERE_HOME/tomcat/conf to Server "Configuration Path".
This can be done inside or outside Eclipse.
The files concerned are: _catalina.policy, catalina.properties, context.xml, jaspic-providers.xml, jaspic-providers.xsd, logging.properties, server.xml, tomcat-users.xml, tomcat-users.xsd and web.xml_. 
![Selection_517](https://user-images.githubusercontent.com/1789408/215547636-c94aec42-47ea-47af-9995-be770a1d2f8f.png)

The "Configuration Path" from the Eclipse Server Configuration is for example /Servers/Apache Tomcat v9; "Servers" is an Eclipse project.
It seems that only _server.xml_ contains relevant data. The port seems to be important.
![Selection_518](https://user-images.githubusercontent.com/1789408/215547764-f36238ae-336b-4fbb-98f8-802b44ac9407.png)

**Step 6**
In _Servers_, double-click on the recently created server to open the _Server Configuration_.
Make sure to
- define a Timeout long enough.
- The "Tomcat admin port" has to match the Server Port from _Server Configuration/server.xml_.
- HTTP-Port: es may be different from the installation, but it must be free for example with commando **ufw** in Ubuntu.
Most errors come from port configuration, so be careful what you define.
![imagen](https://user-images.githubusercontent.com/1789408/215553923-64aeb2ee-e2ff-43be-9ede-a352234be4e8.png)

**Step 7**
- In _Server Configuration_,  click on _Open launch configuration_
- In tab **_Arguments_**, section **_VM Arguments_**, add the following entry: `-DPropertyFile="/PATH_TO_PROPERTIES_FILE/Adempiere.properties" `
- In this file only the database is important. Make sure the database exists, it is an ADempiere DB and its version is compatible with the Eclipse project ADempiere version.
- You can point to the installation Adempiere.properties file.

![Selection_499](https://user-images.githubusercontent.com/1789408/215548579-d029ddfe-f7a7-49ea-ba60-249bc340cff1.png)

**Step 8**
Still in Server configuration, "Open launch configuration"
In tab "Classpath", "User Entries/Add external Jars"  add the following jar files:
- all beneath _$TOMCAT_INSTALLATION_PATH/bin_
- all beneath _$TOMCAT_INSTALLATION_PATH/lib_
 - all beneath _$ECLIPSE_WORKSPACE/JasperReportsTools/lib_
- all beneath _$ECLIPSE_WORKSPACE/tools/lib_
   Exception: all directories and the following files  _libraries.txt, SoftwareLicenses.xls und zxing-2.3.0-LICENSE.txt_ .
- all beneath _$ECLIPSE_WORKSPACE/zkwebui/WEB-INF/lib_

![Selection_500](https://user-images.githubusercontent.com/1789408/215550629-d565a1d6-b17a-4b0c-b4e6-f613600e806f.png)
![Selection_501](https://user-images.githubusercontent.com/1789408/215550639-395363bd-9c15-4aed-bfa3-5da0c99ae368.png)
![Selection_503](https://user-images.githubusercontent.com/1789408/215550658-18d0141f-5d47-4bc5-8894-c286aa398c57.png)

**Step 9**
In the tab "Classpath", "User Entries/Add Projects" add the ADempiere Project.
![Selection_506](https://user-images.githubusercontent.com/1789408/215552844-05cf407c-ad28-4078-a925-82470f5de78b.png)

**Step 10**
In tab "Environment" add the following entry:
Name:  _ADEMPIERE_APPS_TYPE_ 
Value: _tomcat_
![Selection_508](https://user-images.githubusercontent.com/1789408/215554898-36cdf234-f5e9-4e82-8afa-ce7eaeb1af4a.png)

**Step 11**
Start/Debug Tomcat from Eclipse.
There are different ways.
On way is to select the project, then in context menu _Debug As/Debug on Server_
![Selection_511](https://user-images.githubusercontent.com/1789408/215557499-8ca4446e-9551-4b0b-bfcd-d2f5b167d6e0.png)

Select Project
![Selection_512](https://user-images.githubusercontent.com/1789408/215557669-6b3fca9f-1703-4edf-9ba8-4ff8d79b2daf.png)

Wait project to be loaded and executed. There can be some exceptions, but in the end, the database connection must be established:
![Selection_513](https://user-images.githubusercontent.com/1789408/215557928-dc4f87bf-922e-449c-a50b-1b60a0e01619.png)

**Step 12**
Depending on your operating system, Eclipse shows a tab with ADempiere running, or ADempiere runs in a browser:
![Selection_514](https://user-images.githubusercontent.com/1789408/215558280-ef6e63d0-726b-4949-9622-febbce0e3bb9.png)
![Selection_515](https://user-images.githubusercontent.com/1789408/215558293-82409e7c-6f32-4502-885d-d6172ca5f0d9.png)

See also:
-  https://adempiere.io/docs/develop-guide/getting-started/eclipse-ide.html#creating-tomcat-server
- https://adempiere.io/docs/develop-guide/getting-started/eclipse-ide.html

**Comment by augusto741502** [comment]
Hello, I am trying to set up adempiere developer in eclipse 2024, but it sends me a java version error
arning: [options] location of system modules is not set in conjunction with -source 11
 [javac] not setting the location of system modules may lead to class files that cannot run on JDK 11
 [javac] --release 11 is recommended instead of -source 11 -target 11 because it sets the location of system modules automatically
I have tried to find the corresponding version but I cannot get the project up and running, I would ask you please if you can help me with this.

Greetings
